### DISCUSSION TITLE
Use your own input form

### DISCUSSION THREAD

**Discussion by chuppasik** [serverpod/serverpod#2913] [Q&A] [answered]
How can I create my own login form?

**Comment by SandPod** [comment, marked as answer]
Hello @chuppasik 👋 

I'm assuming you would like to create your own login form with the existing authentication module.

To do this, create your form and interact with the endpoints provided by the authentication module.

The endpoints are defined here: https://github.com/serverpod/serverpod/tree/main/modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints
And can be called in the client through the auth module, `await client.modules.auth.<auth provider>`, where `<auth provider>` is the provider implementation you would like to interact with.

You can inspect the implementation of the existing forms to get ideas for how to interact with the provider's methods to get authentication working.

Here is for example the implementation for the email sign-in form: https://github.com/serverpod/serverpod/tree/main/modules/serverpod_auth/serverpod_auth_email_flutter

And in the same folder structure you can find the implementations of the other providers.

**Comment by chuppasik** [comment]
I can't save the authorization state. I register a user, go through authorization, but it is not saved. After I try to use the endpoint, the console reports error 401, user not authenticated. Although in the console before that it reports that the user is authorized.
I don't understand how I can save user authorization, for example with riverpod?
I don't quite understand why I need Session session parameters.
Do I have to initialize the variable final session =...etc?

**Reply by SandPod** [reply]
Did you instantiate a `sessionManager` and register the user there? https://docs.serverpod.dev/concepts/authentication/providers/custom-providers#client-setup

You can implement your own session manager as well.
But if you inspect the flutter code for the email authentication you can see that we register the user in the `sessionManager` after authentication is complete: https://github.com/serverpod/serverpod/blob/939cb3899311a9eb79f1a567b2d2875e6b6e5b60/modules/serverpod_auth/serverpod_auth_email_flutter/lib/src/auth.dart#L35

The `sessionManager` is responsible for passing the authentication token with all requests sent to the server.

**Reply by chuppasik** [reply]
> Вы создали экземпляр `sessionManager`и зарегистрировали там пользователя? https://docs.serverpod.dev/concepts/authentication/providers/custom-providers#client-setup
> 
> Вы также можете реализовать свой собственный менеджер сеансов. Но если вы проверите код flutter для аутентификации электронной почты, вы увидите, что мы регистрируем пользователя после `sessionManager`завершения аутентификации:
> 
> https://github.com/serverpod/serverpod/blob/939cb3899311a9eb79f1a567b2d2875e6b6e5b60/modules/serverpod_auth/serverpod_auth_email_flutter/lib/src/auth.dart#L35
> 
> Отвечает `sessionManager`за передачу токена аутентификации со всеми запросами, отправляемыми на сервер.

Thanks for your help, I solved this problem.
I have one more question, if you don't mind answering. Is it possible to implement authorization from the server side?
I tried to call the registration methods in the server.dart file itself, but it didn't work. I still couldn't send email/pass to the database.

**Reply by SandPod** [reply]
That should be possible.

If you are using the latest serverpod verison (beta) you can inspect the "next" documentation here, it has been revised to help make it simpler to implement custom authentication using the existing email provider: https://docs.serverpod.dev/next/concepts/authentication/providers/custom-providers
